lib/cusolver/cusolverRF.jl

```diff
@@ -0,0 +1,429 @@
+
+    # Assemble internal data structures
+    cusolverRfSetupHost(
+        n, lu_host.nnzA, lu_host.rowsA, lu_host.colsA, lu_host.valsA,
+        lu_host.nnzL, lu_host.rowsL, lu_host.colsL, lu_host.valsL,
+        lu_host.nnzU, lu_host.rowsU, lu_host.colsU, lu_host.valsU,
+        lu_host.P, lu_host.Q,
+        rf
+    )
+    # Analyze available parallelism
+    CUDA.@sync cusolverRfAnalyze(rf)
+    # LU factorization
+    CUDA.@sync cusolverRfRefactor(rf)
+
+    return RfLU{T}(
+        rf, nrhs, n, m, lu_host.nnzA,
+        lu_host.rowsA, lu_host.colsA, lu_host.P, lu_host.Q, d_T
+    )
+end
+
+# Update factorization inplace
+function rf_refactor!(rflu::RfLU{T}, A::CuSparseMatrixCSR{T}) where T
+    CUDA.@sync cusolverRfResetValues(
+        rflu.n, rflu.nnzA,
+        rflu.drowsA, rflu.dcolsA, A.nzVal, rflu.dP, rflu.dQ,
+        rflu.rf
+    )
+    CUDA.@sync cusolverRfRefactor(rflu.rf)
+    return
+end
+
+# Solve system Ax = b
+function rf_solve!(rflu::RfLU{T}, x::CuVector{T}) where T
+    n = rflu.n
+    cusolverRfSolve(rflu.rf, rflu.dP, rflu.dQ, rflu.nrhs, rflu.dT, n, x, n)
+    return
+end
+
+# Batch factorization should not mix with classical LU factorization.
+# We implement a structure apart.
+struct RfBatchLU{T} <: LinearAlgebra.Factorization{T}
+    rf::RfHandle
+    batchsize::Int
+    n::Int
+    m::Int
+    nnzA::Int
+    drowsA::CuVector{Cint}
+    dcolsA::CuVector{Cint}
+    dP::CuVector{Cint}
+    dQ::CuVector{Cint}
+    dT::CuVector{T}
+end
+
+function RfBatchLU(
+    A::CuSparseMatrixCSR{T}, batchsize::Int;
+    fast_mode=true, ordering=:AMD,
+) where T
+    n, m = size(A)
+    lu_host = RfHostLU(A; ordering=ordering)
+
+    # Allocations (device)
+    d_T = CUDA.zeros(Cdouble, m * batchsize * 2)
+
+    rf = RfHandle(; fast_mode=fast_mode)
+
+    # Assemble internal data structures
+    h_valsA_batch = Vector{Float64}[lu_host.valsA for i in 1:batchsize]
+    ptrA_batch = pointer.(h_valsA_batch)
+    cusolverRfBatchSetupHost(
+        batchsize,
+        n, lu_host.nnzA, lu_host.rowsA, lu_host.colsA, ptrA_batch,
+        lu_host.nnzL, lu_host.rowsL, lu_host.colsL, lu_host.valsL,
+        lu_host.nnzU, lu_host.rowsU, lu_host.colsU, lu_host.valsU,
+        lu_host.P, lu_host.Q,
+        rf,
+    )
+    # Analyze available parallelism
+    CUDA.@sync cusolverRfBatchAnalyze(rf)
+    # LU factorization
+    CUDA.@sync cusolverRfBatchRefactor(rf)
+
+    return RfBatchLU{T}(
+        rf, batchsize, n, m, lu_host.nnzA,
+        lu_host.rowsA, lu_host.colsA, lu_host.P, lu_host.Q, d_T
+    )
+end
+
+# Update factorization inplace
+## Single matrix
+function rf_batch_refactor!(rflu::RfBatchLU{T}, A::CuSparseMatrixCSR{T}) where T
+    ptrs = [pointer(A.nzVal) for i in 1:rflu.batchsize]
+    Aptrs = CuArray(ptrs)
+    CUDA.@sync cusolverRfBatchResetValues(
+        rflu.batchsize, rflu.n, rflu.nnzA,
+        rflu.drowsA, rflu.dcolsA, Aptrs, rflu.dP, rflu.dQ,
+        rflu.rf
+    )
+    CUDA.@sync cusolverRfBatchRefactor(rflu.rf)
```

Comment:
Why the sync?

Also, `Aptrs` is now unused, so you should `unsafe_free!` it (to alleviate GC pressure a bit).

Reply:
Good remark for `unsafe_free!`. I have included them in the rescoped version.